### PR TITLE
AWS: Include IPv6 addresses in NodeAddresses

### PR DESCRIPTION
**Pull request by dims**
The change modifies the in-tree cloud provider code to append ipv6 addresses to the Node object advertised by the kubelet.

Signed-off-by: Jyoti Mahapatra<jyotima@amazon.com>
Signed-off-by: Davanum Srinivas <davanum@gmail.com>

#### What type of PR is this?

/kind bug
/kind cleanup

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
Closed the previous PR in this space https://github.com/kubernetes/kubernetes/pull/86918

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go [hunk 1]
```diff
@@ -1610,6 +1644,25 @@ func extractNodeAddresses(instance *ec2.Instance) ([]v1.NodeAddress, error) {
...
+	// IPv6. Ordered after IPv4 addresses, so legacy code can continue to just use the "first" address.
+	for _, networkInterface := range instance.NetworkInterfaces {
+		// skip network interfaces that are not currently in use
+		if aws.StringValue(networkInterface.Status) != ec2.NetworkInterfaceStatusInUse {
+			continue
+		}
+
+		for _, addr6 := range networkInterface.Ipv6Addresses {
+			if ipAddress := aws.StringValue(addr6.Ipv6Address); ipAddress != "" {
+				ip := net.ParseIP(ipAddress)
```
Comment (aojea) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go:1656]: the linter will complain here, you have to use the utilnet.ParseIPSloppy parses

Comment (dims) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: thanks @aojea 

fixed!

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go [hunk 2]
```diff
@@ -1521,6 +1533,27 @@ func (c *Cloud) NodeAddresses(ctx context.Context, name types.NodeName) ([]v1.No
 			}
 		}
 
+		// IPv6. Ordered after IPv4 addresses, so legacy code can continue to just use the "first" address in a dual-stack cluster.
+		for _, macID := range macIDs {
+			ipPath := path.Join("network/interfaces/macs/", macID, "ipv6s")
+			ips, err := c.metadata.GetMetadata(ipPath)
+			if err != nil {
+				if isAWSNotFound(err) {
```
Comment (aojea) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go:1541]: L1487 has other request to metadata but you don't check for the not found error, is this specific of ipv6?

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go [hunk 3]
```diff
@@ -1521,6 +1533,27 @@ func (c *Cloud) NodeAddresses(ctx context.Context, name types.NodeName) ([]v1.No
...
+					// No IPv6 configured. Not an error, just a disappointment.
+					continue
+				}
+				return nil, fmt.Errorf("error querying AWS metadata for %q: %q", ipPath, err)
+			}
+
+			for _, ip := range strings.Split(ips, "\n") {
+				if ip == "" {
```
Comment (aojea) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go:1549]: I remember to discuss with one person from AWS before about IPs in non canonical forms, should we canonicalize the IPv6 address?

I can try to find the discussion later if that can help

Comment (aojea) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go:1549]: canonicalize means parse the ip and output the string format from the golang stdlib `netutils.ParseIPSloppy(ip).String()`

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go [hunk 4]
```diff
@@ -1610,6 +1643,25 @@ func extractNodeAddresses(instance *ec2.Instance) ([]v1.NodeAddress, error) {
...
+		}
+
+		for _, addr6 := range networkInterface.Ipv6Addresses {
+			if ipAddress := aws.StringValue(addr6.Ipv6Address); ipAddress != "" {
+				ip := netutils.ParseIPSloppy(ipAddress)
+				if ip == nil {
+					return nil, fmt.Errorf("EC2 instance had invalid IPv6 address: %s (%q)", aws.StringValue(instance.InstanceId), ipAddress)
+				}
+				// NB: "Internal" is actually about intra-cluster reachability, and not public vs private.
+				addresses = append(addresses, v1.NodeAddress{Type: v1.NodeInternalIP, Address: ip.String()})
```
Comment (aojea) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go:1660]: ok, I see, I suggested do the same that you are doing here in https://github.com/kubernetes/kubernetes/pull/113114/files#r998265747 
but better ack from one with more knowledge in this provider

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go (hunk 3 above)
Comment (aojea) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go:1549]: oh, wait, forget it, I think that the problem is if you canonicalize this is going to hotloop, scratch this comment

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go [hunk 5]
```diff
@@ -1610,6 +1643,25 @@ func extractNodeAddresses(instance *ec2.Instance) ([]v1.NodeAddress, error) {
...
+			continue
+		}
+
+		for _, addr6 := range networkInterface.Ipv6Addresses {
+			if ipAddress := aws.StringValue(addr6.Ipv6Address); ipAddress != "" {
+				ip := netutils.ParseIPSloppy(ipAddress)
+				if ip == nil {
+					return nil, fmt.Errorf("EC2 instance had invalid IPv6 address: %s (%q)", aws.StringValue(instance.InstanceId), ipAddress)
+				}
+				// NB: "Internal" is actually about intra-cluster reachability, and not public vs private.
```
Comment (aojea) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go:1659]: is aws returning link local addresses, do you want to filter here to obtain only reachable addresses?

**Comment by andrewsykim** [comment]
I think the original PR was not merged because of the "no new features for in-tree providers" policy we put in place since v1.20. Which is why IPv6 efforts were shifted to `k8s.io/cloud-provider-aws` and EKS distro.

**Comment by dims** [comment]
@andrewsykim ACK, let me close this out then

**Pull request closed without merging**